crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,1039 @@
+
+    fn accept<T>(&self, visitor: &mut T)
+    where
+        T: DataTypeVisitor<T>,
+    {
+        visitor.visit(self);
+    }
+
+    fn not_null(&self) -> Self {
+        self.copy(false)
+    }
+
+    fn nullable(&self) -> Self {
+        self.copy(true)
+    }
+}
+
+/// ArrayType for paimon.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/ArrayType.java>.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct ArrayType {
+    pub element_type: DataType,
+}
+
+impl Display for ArrayType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if !self.element_type.is_nullable() {
+            write!(f, "ARRAY NOT NULL")
+        } else {
+            write!(f, "ARRAY")
+        }
+    }
+}
+
+impl ArrayType {
+    pub fn new(is_nullable: bool) -> Self {
+        Self {
+            element_type: DataType::new(is_nullable, DataTypeRoot::Array),
+        }
+    }
+
+    pub fn default_value() -> Self {
+        Self::new(true)
+    }
+}
+
+/// BigIntType for paimon.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/master/paimon-common/src/main/java/org/apache/paimon/types/BigIntType.java>.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Hash)]
+pub struct BigIntType {
+    pub element_type: DataType,
+}
+
+impl Display for BigIntType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if !self.element_type.is_nullable() {
+            write!(f, "BIGINT NOT NULL")
+        } else {
+            write!(f, "BIGINT")
+        }
+    }
+}
+
+impl BigIntType {
+    pub fn new(is_nullable: bool) -> Self {
+        Self {
+            element_type: DataType::new(is_nullable, DataTypeRoot::Bigint),
+        }
+    }
+
+    pub fn default_value() -> Self {
+        Self::new(true)
+    }
+}
+
+/// BinaryType for paimon.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/master/paimon-common/src/main/java/org/apache/paimon/types/BinaryType.java>.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct BinaryType {
+    pub element_type: DataType,
+    length: usize,
+}
+
+impl Display for BinaryType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if !self.element_type.is_nullable() {
+            write!(f, "BINARY({}) NOT NULL", self.length)
+        } else {
+            write!(f, "BINARY({})", self.length)
+        }
+    }
+}
+
+impl BinaryType {
+    pub const MIN_LENGTH: usize = 1;
+
+    pub const MAX_LENGTH: usize = isize::MAX as usize;
+
+    pub const DEFAULT_LENGTH: usize = 1;
+
+    pub fn new(is_nullable: bool, length: usize) -> Self {
+        Self::new_with_result(is_nullable, length).unwrap()
+    }
+
+    pub fn new_with_result(is_nullable: bool, length: usize) -> Result<Self, &'static str> {
```

Comment:
We can call this API `fn try_new() -> Result<Self, Error>`. 

The Error type is not defined yet, I will polish this part later.